src/core/lv2/Lv2SubPluginFeatures.cpp

```diff
@@ -0,0 +1,186 @@
+{
+	const LilvPlugin* result = Engine::getLv2Manager()->
+		getPlugin(k.attributes["uri"]);
+	Q_ASSERT(result);
+	return result;
+}
+
+
+
+
+QString Lv2SubPluginFeatures::pluginName(const LilvPlugin *plug)
+{
+	return qStringFromPluginNode(plug, lilv_plugin_get_name);
+}
+
+
+
+
+Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::PluginTypes type) :
+	SubPluginFeatures(type)
+{
+}
+
+
+
+
+void Lv2SubPluginFeatures::fillDescriptionWidget(QWidget *parent,
+													const Key *k) const
+{
+	const LilvPlugin *plug = getPlugin(*k);
+
+	QLabel *label = new QLabel(parent);
+	label->setText(QWidget::tr("Name: ") + pluginName(plug));
+
+	QLabel *label2 = new QLabel(parent);
+	label2->setText(QWidget::tr("URI: ") +
+		lilv_node_as_uri(lilv_plugin_get_uri(plug)));
+
+	QWidget *maker = new QWidget(parent);
+	QHBoxLayout *l = new QHBoxLayout(maker);
+	l->setMargin(0);
+	l->setSpacing(0);
+
+	QLabel *maker_label = new QLabel(maker);
+	maker_label->setText(QWidget::tr("Maker: "));
+	maker_label->setAlignment(Qt::AlignTop);
+
+	QLabel *maker_content = new QLabel(maker);
+	maker_content->setText(
+		qStringFromPluginNode(plug, lilv_plugin_get_author_name));
+	maker_content->setWordWrap(true);
+
+	l->addWidget(maker_label);
+	l->addWidget(maker_content, 1);
+
+	QWidget *copyright = new QWidget(parent);
+	l = new QHBoxLayout(copyright);
+	l->setMargin(0);
+	l->setSpacing(0);
+	copyright->setMinimumWidth(parent->minimumWidth());
+
+	QLabel *copyright_label = new QLabel(copyright);
+	copyright_label->setText(QWidget::tr("Copyright: "));
+	copyright_label->setAlignment(Qt::AlignTop);
+
+	QLabel *copyright_content = new QLabel(copyright);
+	copyright_content->setText("<unknown>");
+	copyright_content->setWordWrap(true);
+	l->addWidget(copyright_label);
+	l->addWidget(copyright_content, 1);
+
+	AutoLilvNodes extensions(lilv_plugin_get_extension_data(plug));
+	(void)extensions;
+	// possibly TODO: version, project, plugin type, number of channels
+}
+
+
+
+
+QString Lv2SubPluginFeatures::additionalFileExtensions(
+	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
+{
+	(void)k;
+	// lv2 only loads .lv2 files
+	// maybe add conversions later, e.g. for loading xmz
+	return QString();
+}
+
+
+
+
+QString Lv2SubPluginFeatures::displayName(
+	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
+{
+	return pluginName(getPlugin(k));
+}
+
+
+
+
+QString Lv2SubPluginFeatures::description(
+	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
+{
+	(void)k;
+	return QString::fromUtf8("description not implemented yet"); // TODO
+}
+
+
+
+
+const PixmapLoader *Lv2SubPluginFeatures::logo(
+	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
+{
+	(void)k; // TODO
+	return nullptr;
+}
+
+
+
+
+void Lv2SubPluginFeatures::listSubPluginKeys(const Plugin::Descriptor *desc,
+												KeyList &kl) const
+{
+	Lv2Manager *lv2Mgr = Engine::getLv2Manager();
+	for (const std::pair<const std::string, Lv2Manager::Lv2Info> &pr :
+		*lv2Mgr)
```

Comment:
I prefer the non auto sometimes as without the info, a user might be wondering "what is `pr.second`"? With a good IDE, you see the type fast, but e.g. here in the web browser... I'm personally undecided. Does the argument with the readability bias your decision?

Reply:
It might help if you can name `first` and `second`. You miss types a lot less than you expect, but in the absence of a useful name, sometimes they're all you've got to go off.

In C++17 you can use a structured binding:
```cpp
for(const auto &[uri, info] : *lv2Mgr) {
	// ...
}
```
but I'm not sure if we can use that here yet.
You could name them manually within the loop body:
```cpp
for (const auto &pr : *lv2Mgr) {
	const auto &uri = pr.first;
	const auto &info = pr.second;
	// ...
}
```
but this adds two more lines of code - I'll let you weigh that up.
Finally, you could just rename the loop variable to something more explanatory:
```cpp
for(const auto &uriInfoPair : *lv2Mgr) {
	// ...
}
```

Reply:
As usual, I prefer the idea you list as "finally" :smile: